bindings/python/src/utils/normalization.rs

```diff
@@ -0,0 +1,504 @@
+    // TODO: Add the compatibility for Fn(char) -> bool
+}
+
+impl Pattern for PyPattern<'_> {
+    fn find_matches(&self, inside: &str) -> tk::Result<Vec<(tk::Offsets, bool)>> {
+        match self {
+            PyPattern::Str(s) => {
+                let mut chars = s.chars();
+                if let (Some(c), None) = (chars.next(), chars.next()) {
+                    c.find_matches(inside)
+                } else {
+                    s.find_matches(inside)
+                }
+            }
+            PyPattern::Regex(r) => {
+                Python::with_gil(|py| (&r.borrow(py).inner).find_matches(inside))
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone, FromPyObject)]
+pub enum PyRange<'s> {
+    #[pyo3(annotation = "int")]
+    Single(isize),
+    #[pyo3(annotation = "Tuple[uint, uint]")]
+    Range(usize, usize),
+    #[pyo3(annotation = "slice")]
+    Slice(&'s PySlice),
+}
+impl PyRange<'_> {
+    pub fn to_range(&self, max_len: usize) -> PyResult<std::ops::Range<usize>> {
+        match self {
+            PyRange::Single(i) => {
+                if i.is_negative() {
+                    let i = -i as usize;
+                    if i > max_len {
+                        Err(exceptions::PyValueError::new_err(format!(
+                            "{} is bigger than max len",
+                            i
+                        )))
+                    } else {
+                        Ok(max_len - i..max_len - i + 1)
+                    }
+                } else {
+                    let i = *i as usize;
+                    Ok(i..i + 1)
+                }
+            }
+            PyRange::Range(s, e) => Ok(*s..*e),
+            PyRange::Slice(s) => {
+                let r = s.indices(max_len as std::os::raw::c_long)?;
+                Ok(r.start as usize..r.stop as usize)
+            }
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct PySplitDelimiterBehavior(SplitDelimiterBehavior);
+
+impl FromPyObject<'_> for PySplitDelimiterBehavior {
+    fn extract(obj: &PyAny) -> PyResult<Self> {
+        let s = obj.extract::<&str>()?;
+
+        Ok(Self(match s {
+            "removed" => Ok(SplitDelimiterBehavior::Removed),
+            "isolated" => Ok(SplitDelimiterBehavior::Isolated),
+            "merged_with_previous" => Ok(SplitDelimiterBehavior::MergedWithPrevious),
+            "merged_with_next" => Ok(SplitDelimiterBehavior::MergedWithNext),
+            "contiguous" => Ok(SplitDelimiterBehavior::Contiguous),
+            _ => Err(exceptions::PyValueError::new_err(
+                "Wrong value for SplitDelimiterBehavior, expected one of: \
+                `removed, isolated, merged_with_previous, merged_with_next, contiguous`",
+            )),
+        }?))
+    }
+}
+
+impl From<PySplitDelimiterBehavior> for SplitDelimiterBehavior {
+    fn from(v: PySplitDelimiterBehavior) -> Self {
+        v.0
+    }
+}
+
+fn filter(normalized: &mut NormalizedString, func: &PyAny) -> PyResult<()> {
```

Comment:
Again that seems a bit too much api here. one very generic function seems better to lower the surface we have to maintain. Rest of the helper functions can be in pure python I feel.

Reply:
Nothing can be in pure python unfortunately, for normalization, we have to use the `NormalizedString`. Its goals is to keep track of the alignment, while pure python doesn't. This is just another method on `NormalizedString`.